### DISCUSSION TITLE
Reactions to nested Atoms

### DISCUSSION THREAD

**Discussion by slimshader** [codewriter-packages/UniMob#7] [Q&A] [answered]
Hi,

I think I am missing something still, code sample:

```
public class ExperimentsWindow : Window
{
    class ViewModel
    {
        [Atom]
        public int IntValue { get; set; } 
    }

    [Atom]
    ViewModel VM { get; set; }

    private void Start()
    {
        Initialize();

        VM = new ViewModel() { IntValue = 5 };

        // tried this too: Atom.Reaction(() => Debug.Log($"Int value {VM.IntValue}"));
        Atom.Reaction(() => VM.IntValue, i => Debug.Log($"Int value {i}"));

        VM.IntValue = 6;

        VM = new ViewModel() { IntValue = 7 };
    }
}
```

When running, I always only see values "5" or "7" printed but not "6". Is there a way to React to changes in multi-level Atoms?

**Comment by vanifatovvlad** [comment, marked as answer]
The reactions are executed in two cases:
1) on creation;
2) on the next frame after any dependant atom changed.

In this sample, a value "5" is printed immediately when creating a reaction. Values "6" and "7" are set during one frame. They trigger a reaction, but the reaction itself is executed once on the next frame and use latest values (so "7" is printed).

The value "6" will be printed if the values are set in different frames, or if you force actualize a reaction.

```csharp
private IEnumerator Start()
{
    VM = new ViewModel() {IntValue = 5};

    Atom.Reaction(() => Debug.Log($"Int value {VM.IntValue}"));

    VM.IntValue = 6;
    yield return null; // skip one frame

    VM = new ViewModel() {IntValue = 7};
}
```

```csharp
private void Start()
{
    VM = new ViewModel() {IntValue = 5};

    var reaction = Atom.Reaction(() => Debug.Log($"Int value {VM.IntValue}"));

    VM.IntValue = 6;
    reaction.Activate(); // force actualize

    VM = new ViewModel() {IntValue = 7};
}
```

**Comment by slimshader** [comment]
Ah, makes perfect sense now. The fact that Reaction is only triggered next frame with latest value could be leveraged as optimization I suppose then, no need to discard intermediate updates to the values as they will not force UI update anyways.

Question: could you clarify the difference between Reaction(Action reaction) and Reaction(AtomPull react, Action effect)? When to use which one?

**Comment by vanifatovvlad** [comment]
_Reaction(AtomPull react, Action effect)_ tracks atoms that used in _react_ method, and won't track atoms in _effect_ method. 

For example, this reaction:
```csharp
Atom.Reaction(() => {
   var val = myAtom;
   otherPlugin.Process(val);
});
```
will executed when myAtom value changed, but potentially otherPlugin can access other atoms in Process method, so reaction will track them too.

If you dont want this behaviour, you can use _Reaction(AtomPull react, Action effect)_ overload to control which data should be tracked.

In this sample reaction wont track atoms used in Process method.
```csharp
Atom.Reaction(
   () => myAtom,
   val => otherPlugin.Process(val)
);
```

Or you can use NoWatch
```csharp
Atom.Reaction(() => {
   var val = myAtom;
   using (Atom.NoWatch) {
      otherPlugin.Process(val);
  }
});
```
